min-turnup: azure

colemickens opened the pull request:
I changed:
- `--service-cluster-ip-range` to not overlap with the ip range that will be used for the actual nodes.
- changed the manifest templates to jinja2, it was easier to do the syntax I needed for adding the `--cloud-config` flag
- changed the master pod manifests to use `hyperkube-amd64` instead of the `kube-{apiserver,scheduler,controller-manager}` images. `kubelet.service` was already using `hyperkube-amd64` so it's already on box, plus it makes it easier for me to test changes to cloudprovider if I only have to rebuild and upload hyperkube instead of 4+ containers.
- added a `Dockerfile` that has `terraform`, `jq`, `jsonnet`, etc. (especially since this relies of a pre-release version of terraform [https://github.com/hashicorp/terraform/issues/6803]).

This worksaround Azure's lack of metadata service by using terraform's `template_file` functionality. I use terraform's `base64encode` and `file` functions to access the contents of the crypto assets and use `template_file` to interpolate them into a `configure-vm.sh` template that is rendered and passed as `custom_data` to the VM. Similarly it passes through some environment information and Active Directory credentials that are required by the `cloudprovider` implementation. The `template_file` approach was chosen instead of using the `remote-exec` and `file` provisioners in terraform because they don't work easily for Azure currently: https://github.com/hashicorp/terraform/issues/7122.

This relies on the `azure` `cloudprovider`, so it current defaults to my own `hyperkube-amd` repo/image. (I'm blocked on sending the PR for the `cloudprovider` for reasons... but hopefully that will stop being a thing soon)

It should probably include a script to help create the Azure Service Principal that could then be used by Terraform and passed through for `cloudprovider`.

So, it's not ready to merge yet for reasons mentioned above, but I wanted to get this out for any comments. I was able to deploy a 20 cluster on the first try after getting the single master/node working: http://i.imgur.com/boyFZsw.png.

The biggest issue I see is that `addons` seem to be unaddressed. For now I haven't added `kube-proxy` as a manifest pod since I assume we're going to run it via a `DaemonSet` addon....

colemickens committed:
s/conig/config/g

colemickens committed:
min-turnup: azure

colemickens commented:
Other changes:

* added /etc/kubernetes as a volume for the manifest pods (azure cloudprovider config lives at `/etc/kubernetes/azure.json`).

colemickens commented:
Another consideration: I don't think either the gce/azure variant output a user-friendly kubeconfig at the end for connectivity outside of the cluster.

colemickens commented:
for reference, here's a kube-proxy daemonset yaml that works with this: https://github.com/colemickens/polykube/blob/e5e6f463e0aae8e4dee204f5496a066f932f438b/kubernetes/kube-proxy.yaml

mikedanese commented:
Thanks Cole! This looks great, I will give this a review soon. cc @errordeveloper

colemickens commented:
(Another random question I've had -- is it completely safe to run kubelet in container now with Docker 1.11.x? Like volume/secret mounts will work?)

errordeveloper commented:
Secrets work for sure. Formal validation is still required, but it
generally works with 1.10 and 1.11, and secrets are broken with 1.9 and
older versions.

On Tue, 14 Jun 2016, 00:47 Cole Mickens, <notifications@github.com> wrote:

> (Another random question I've had -- is it completely safe to run kubelet
> in container now with Docker 1.11.x? Like volume/secret mounts will work?)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kube-deploy/pull/106#issuecomment-225741249>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAPWS1Dd1Qk9n0ahvj2VpkC_-rvjLCIOks5qLewQgaJpZM4IzsbA>
> .
>

colemickens commented:
(I guess I should've known as much -- I deployed my test app using secrets to the cluster I brought up with this.)

colemickens committed:
fix flaky cluster by letting etcd persist. Add kubernetes service ip as a SAN. Let kubelet restart. Add set-kubeconfig helper for configuring kubeconfig after deployment. Add note about extra-sans-ip being unused.

colemickens commented:
Flakiness is fixed as far as I can tell - etcd had restarted, it hadn't been persisting anywhere on the host and the apiserver/kubelets did not get along. I suspect that kubelets weren't completely re-registering and so were just failing to update their status since the apiserver didn't know about them. Either way, etcd is persisting now and I haven't had problems with nodes disappearing anymore.

I also had to add the cluster ip for the kubernetes service as a SAN so that the certs would be valid for in-cluster-config usages (kubernetes-dashboard was complaining. I don't know if it pre-resolves the kubernetes service clusterip or if it's made to not rely on dns being available. Regardless...)

Added a helper for configuring kubectl after deployment. Probably should be temporary or should be improved as it requires the user to manually get the MASTER_IP and set CLUSTER_NAME even those things can be acquired through Terraform and/or inspecting `.config.json` and reading the instance prefix as a CLUSTER_NAME.

mikedanese commented on min-turnup/phase2/ansible/playbooks/roles/master/templates/kube-apiserver.json.j2:
```diff
@@ -0,0 +1,102 @@
+              "--etcd-servers=http://127.0.0.1:2379",
+              "--cloud-provider={{ phase1['cloud_provider'] }}",
+{% if phase1['cloud_provider'] == "azure" %}
+              "--cloud-config=/etc/kubernetes/azure.json",
```
For comparison, this could be done with jsonnet like this:

``` lua
local build_params(arr) =
  std.flattenArrays(std.filter(function(a) a != null, arr));

build_params([
  [
    "--this",
    "--that",
  ],
  if "gce" == "gce" then
    ["--life"],
  if "azure" == "gce" then
    ["--death"]
  else
    ["--parties"],
])
```

produces:

``` json
[
   "--this",
   "--that",
   "--life",
   "--parties"
]
```

colemickens replied:
Hm, do you think that is better? I think the j2 version is much easier to parse/grok at a glance, albeit with slightly more json cruft.

mikedanese replied:
I want to avoid a templating language with a python dependency because of https://github.com/kubernetes/kubernetes/issues/27547. I prefer to use a language with an interpreter that can be statically linked, e.g. go templates or jsonnet. I don't like using text templating langauges to template data formats so I lean towards jsonnet.

colemickens replied:
Isn't this running from within the ansible container anyway? I'm fine with changing it, but I think it's orthogonal to whether or not python is on the host, no?

mikedanese replied:
It's not so much that we can't possible get it to run on these os'es. It's more about the size of the dependency. python:2.7-slim was at 200MB when I looked. This is larger then the GCI vm image. This hits us on time to boot and is prohibitive. So now we have code like this https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/configure-helper.sh#L813-L845. There's a benifit to sharing templates of some form for the kubernetes components but I don't think that it's not long term viable to do this with jinja. In comparison jsonnet is a couple of MBs and only has a dependency on libc (when it's not statically compiled). Go templates would be similarly lean.

colemickens replied:
Sounds good, I've also noticed that pulling the install-k8s container adds noticeable time to the deployment.

I'll convert it back to jsonnet in my `kubernetes-anywhere` branch.

mikedanese commented:
So we were using /var/etcd instead of /var/lib/etcd before this change which caused data to not be saved between etcd container restarts?

mikedanese commented on min-turnup/phase1/azure/lib/azure.jsonnet:
```diff
@@ -0,0 +1,285 @@
+								DNS.4 = kubernetes.default.svc.cluster.local
+								DNS.5 = names.master_vm
+								IP.1 = ${azurerm_public_ip.pip.ip_address}
+								IP.2 = 10.3.0.1
```
We should probably make node cidr, service cidr, pod cidr configurable ASAP.

errordeveloper replied:
+1

On Tue, 14 Jun 2016, 22:26 Mike Danese, notifications@github.com wrote:

> In min-turnup/phase1/azure/lib/azure.jsonnet
> https://github.com/kubernetes/kube-deploy/pull/106#discussion_r67060401:
> 
> > -               pip: {
> > -                   resource_group_name: "${azurerm_resource_group.rg.name}",
> > -                   location: cfg.azure.location,
> > -                   name: names.master_public_ip,
> > -                   public_ip_address_allocation: "static",
> > -                   provisioner:  [{
> > -                       "local-exec": {
> > -                           command: |||
> > -                               cat <<EOF > ../../phase1b/crypto/san-extras
> > -                               DNS.1 = kubernetes
> > -                               DNS.2 = kubernetes.default
> > -                               DNS.3 = kubernetes.default.svc
> > -                               DNS.4 = kubernetes.default.svc.cluster.local
> > -                               DNS.5 = names.master_vm
> > -                               IP.1 = ${azurerm_public_ip.pip.ip_address}
> > -                               IP.2 = 10.3.0.1
> 
> We should probably make node cidr, service cidr, pod cidr configurable
> ASAP.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/kubernetes/kube-deploy/pull/106/files/f56f5e25dce862b13c240daede4900fe2e9c680e#r67060401,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAPWS4rcn0N4jZ2HvDAL4Jnapaf7vQnvks5qLxx7gaJpZM4IzsbA
> .

errordeveloper replied:
<strike>It seems like a combination of template and file provisioner could probably be better here... </strike> (Please ignore, that won't work for various reasons.) Anyhow, it'd be great to investigate if TLS provider would do the job (see #35).

mikedanese replied:
Using terraform to build the certs SGTM

colemickens replied:
It sounds good to me, but I don't want to try to add that on top of this PR. I can potentially look at in the future (but that would be after azure-cloudprovider).

mikedanese commented on min-turnup/phase2/ansible/playbooks/roles/node/templates/kubelet.service.j2:
```diff
@@ -15,27 +15,31 @@ ExecStart=/usr/bin/docker run \
-{% if role == "master" %}
-            --api-servers=http://localhost:8080 \
-{% elif role == "node" %}
+            --register-node=true \
```
I think this is the default so no need to provide?

mikedanese commented on min-turnup/phase2/ansible/playbooks/roles/node/templates/kubelet.service.j2:
```diff
@@ -15,27 +15,31 @@ ExecStart=/usr/bin/docker run \
-            --hairpin-mode=promiscuous-bridge \
             --kubeconfig=/srv/kubernetes/kubeconfig.json \
+{% if role == "master" %}
+            --config=/etc/kubernetes/manifests \
```
It seems reasonable to enable this on nodes too.

colemickens replied:
Agreed.

mikedanese commented on min-turnup/phase2/ansible/playbooks/roles/node/templates/kubelet.service.j2:
```diff
@@ -15,27 +15,31 @@ ExecStart=/usr/bin/docker run \
+            --cluster-dns=10.3.0.10 \
             --cluster-domain=cluster.local \
-            --v=2
+            --v=9
```
why do you want this?

colemickens replied:
Was debugging my cloudprovider. I'll revert to `2`.

mikedanese commented on min-turnup/phase2/ansible/playbooks/roles/master/templates/kubeconfig.jsonnet:
```diff
@@ -1,26 +0,0 @@
-function(cfg)
```
isn't this required to create /srv/kubernetes/kubeconfig.json? I don't see what it was replaced by.

colemickens replied:
I have no idea why this is showing up like this. In both the upstream branch and my own branch, the file `kubeconfig.jsonnet` lives at `roles/node/templates`. I can't explain why GitHub is showing this diff...

colemickens replied:
Oh, it looks like you had it in both `master`/`node` and I removed the redundant duplicate from `master`.

mikedanese commented on min-turnup/phase1/azure/lib/azure.jsonnet:
```diff
@@ -0,0 +1,285 @@
+		},
+		resource: {
+			azurerm_resource_group: {
+				rg: {
```
nit, these short abbreviations are pretty hard to decipher.

errordeveloper replied:
True, but `a_thing: { the_thing: { ... } }` doesn't look too much better. Is there a jsonnet trick for such things?

mikedanese replied:
Is it a safe assumption that you always want this map to be keyed by the "name" field of the resource object?

``` lua
local array_to_map(arr, keyfunc) =
  { [keyfunc(e)]: e for e in arr };
local name_keyfunc(e) = e.name;
local resouce_map(arr) = array_to_map(arr, name_keyfunc);

{
  resources: resouce_map([
    {
      name: "resouce-a",
    },
    {
      name: "resouce-b",
    },
  ]),
}
```

produces:

``` json
{
   "resources": {
      "resouce-a": {
         "name": "resouce-a"
      },
      "resouce-b": {
         "name": "resouce-b"
      }
   }
}
```

colemickens replied:
So imagine a jsonnet input for the `azurerm_resource_group` resource would look like:

```
{
  azurerm_resource_group: resource_map([
    {
      name: cfg.instance_prefix,
    },
  ])
}
```

Later on, since all resource wind up "parented" to a resource group, I have to do:

```
{
  azurerm_blahblah_resource: resource_map([
    {
      resource_group_name: "${azurerm_resource_group." + cfg.instance_prefix + ".name}"
    },
  ])
}
```

I don't want to cheat and just use `"resource_group_name": cfg.instance_prefix` because it creates a non-explicit dependency; aka things would break if someone ever modified the Terraform template to append `-rg` in the `azurerm_resource_group.?.name` field. That's sort of why I chose the shorthand names in the first place.

That having been said, that syntax isn't really a big deal. The suggested change is probably more discoverable for someone looking at their terraform output, with what I have now they might think "`azurerm_resource_group.rg`... what is `rg`? My resource group is named `cfg.instance_prefix`".

After typing all of this, I've sort of convinced myself to go with the suggestion. Let me know if you guys have other thoughts.

colemickens replied:
Another note: this doesn't work for storage accounts because I have to strip `-` out of the `instance_prefix`. I do that with Terraform syntax since [jsonnet doesn't support string replace](https://github.com/google/jsonnet/issues/102). With the `resource_map` technique, the whole string with terraform syntax in it becomes the `key` and the escaping doesn't work out.

It renders like this, and if I try to escape it with jsonnet then it won't be correctly parsed by terraform.

``` json
{ "${replace("colemick-kube3strg", "-", "")}": "" }
```

~~I'll just do the storage account resource by hand I guess.~~ Well, it's the same problem if I try to use the real storage account name as the key... it's just going to have to stay as a generic "storage_account" name. I'm starting to wonder if this is worth the effort.

mikedanese commented on min-turnup/phase1/azure/configure-vm.sh:
```diff
@@ -0,0 +1,69 @@
+
+curl -sSL https://get.docker.com/ | sh
+
+apt-get update
```
nit: get.docker.com will do an apt-get update

colemickens replied:
Removed.

mikedanese commented on min-turnup/phase1/azure/configure-vm.sh:
```diff
@@ -0,0 +1,69 @@
+EOF
+cat <<EOF > /etc/systemd/system/docker.service.d/overlay.conf
+[Service]
+ExecStart=
```
I am no systemd expert. Why is this line necessary?

errordeveloper replied:
Well spotted ;) This resets the previous value. Well, it's a bit of a special case for `ExecStart` and certain other attributes, but that's what you have to do and it complains otherwise.

mikedanese commented on min-turnup/phase1/azure/configure-vm.sh:
```diff
@@ -0,0 +1,69 @@
+azure_file="/etc/kubernetes/azure.json"
+config_file="/etc/kubernetes/k8s_config.json"
+
+mkdir /etc/kubernetes
```
mkdir /etc/kubernetes will fail if the directory already exists. Is this script supposed to be idempotent?

colemickens replied:
Addressing this in the other comment you left on this file.

colemickens replied:
(I will add `-p` to prevent it from failing.)

mikedanese commented on min-turnup/phase1/azure/configure-vm.sh:
```diff
@@ -0,0 +1,69 @@
+installer_container_spec="$(cat "$config_file" | jq -r '.phase2.installer_container_spec')"
+
+cat << EOF > /etc/kubernetes/install.sh
+systemctl stop docker
```
this script could use a #!/bin/bash. why do you need to restart docker?

colemickens replied:
RE: this and the question above... I had done some hacky things in `configure-vm.sh`. In Azure, the `configure-vm.sh` script didn't seem to be available in plaintext (there might be a base64 copy somewhere). So what I was doing was writing the `install.sh` script out to disk.

So the `configure-vm.sh` dropped the `install.sh` script for future runs, as I was iterating on the Ansible container and then invoked it at the end of `configure-vm.sh` for the initial deployment.

The thinking here was that someone would want to upgrade their cluster. They've updated the `k8s_config.json` to reference the newer Kubernetes version, now they need to re-run the playbook container. With my setup, they just call `install.sh` again. Kubelet gets torn down as part of rebooting Docker. Playbook runs, new kubelet boots up. So `install.sh` is what I was treating as needing to be idempotent.

(and yes, it should have a shebang, if we want to keep this model. I don't know if `configure-vm.sh` is meant to be re-run on GCE?)

errordeveloper commented on min-turnup/phase1/azure/lib/azure.jsonnet:
```diff
@@ -0,0 +1,285 @@
+			},
+			template_file: {
+				azure_json: {
+					template: "${file(\"azure.json\")}",
```
I wonder if this could be rendered with jsonnet instead, once we are using it...?

mikedanese replied:
I agree with this sentiment for files that don't create secrets. If it's possible we should avoid schlepping around secrets everywhere

colemickens replied:
Not sure what you're talking about? This is using terraform's "template_file" functionality. Terraform will interpolate values into the `azure.json` template. The rendered form is then used as custom_data in the VM resource/object.

Using jsonnet would require me to somehow write those values out to a file, then execute jsonnet from within terraform, and then regrab that output to pass as custom_data to the VM. (actually, it gets used by the `configure-vm` `template_file` which is then used as the custom_data.)

It seems like more work for an unknown reward.

errordeveloper commented on min-turnup/phase1/azure/lib/azure.jsonnet:
```diff
@@ -0,0 +1,285 @@
+				configure_vm: {
+					template: "${file(\"configure-vm.sh\")}",
+					vars: {
+						apiserver_tar: "${base64encode(file(\"../../phase1b/crypto/apiserver.tar\"))}",
```
<strike>Is there a specific reason for not using file provisioner for these? The only reasons I can think of are that we might want to avoid depending on SSH access, as some use might want to disable it, or don't want to make Terraform wait for SSH for whatever other reasons...?</strike> Sorry, should have read the description before reading the code.

colemickens replied:
Yes, I mentioned in the PR text. Terraform doesn't have file provisioners working for Azure yet, or else I would have used them.

mikedanese replied:
Terraform file provisioning requires terraform to be running and provisioning the nodes. This seems problematic if we want to be able to resize the minion ASG/MIG and expect new nodes to join the cluster as minions, right?

colemickens replied:
This is true. That is why I shoved everything in customData/cloudInit in my other bring-up as well.

errordeveloper replied:
Yes, something I totally haven't thought of. May be we should make an note in docs or via an issue for it, so it's clear nobody uses those.

errordeveloper commented on min-turnup/phase2/ansible/Makefile:
```diff
@@ -1,5 +1,6 @@
 TAG=v1
-IMAGE=gcr.io/mikedanese-k8s/install-k8s
+#IMAGE=gcr.io/mikedanese-k8s/install-k8s
+IMAGE=docker.io/colemickens/install-k8s
```
Let's set this to `IMAGE ?= bar`, so it's easy to override with `make IMAGE=foo`.

colemickens replied:
👍

errordeveloper commented on min-turnup/phase2/Kconfig:
```diff
@@ -3,14 +3,18 @@ menu "Phase 2: Node Bootstrapping"
 config phase2.kubernetes_version
 	string "kubernetes version"
-	default "v1.2.4"
+	default "v1.3.0-azure"
```
I think this is a local change, isn't it?

colemickens replied:
Yes, but the Azure bring-up will not work as-is without using my branch with the `cloudprovider`. That's one of the caveats I called out that makes this a bit harder to handle.

If it were possible to use flannel (without requiring the docker bootstrap) then we could potentially support Azure + min_turnup without requiring the cloudprovider. I'm not keen on doing that immediately since I still need to get the cloudprovider reviewed/merged first.

I'm open to suggestions on how to handle this.

errordeveloper replied:
> If it were possible to use flannel (without requiring the docker bootstrap)

With Weave Net you don't have to care about second Docker daemon, cause Weave Net doesn't reconfigure the `docker0` bridge, it instead manages it's own bridge independently. Happy to help. Azure route configuration is plausible feature request for Weave Net, I think, but that's a separate topic.

mikedanese commented:
@colemickens what are your thoughts after implementing a phase 1? How was it better than worse then implementing a kube-up.sh? Excluding the rewrite of the manifest files, the diff seems fairly small.

colemickens commented:
Thank you both for the review and comments. They all make sense and I'll address these ASAP, but it may take a few days - I'm out of town right now for my brother's wedding and it's taking more of my attention than I'd expected. (I'm also simultaneously trying to get the cloudprovider itself out for review.)

colemickens committed:
changes for some of initial feedback

colemickens commented:
@mikedanese Sorry, I muddled the etcd/flakiness discussion, sorry. I had removed the hostPath volume entirely for Azure since it was using the GCE PD mount path. The fix was that I re-added hostPath, not that I changed the path (that was unintentional). I've actually reverted the path to `/var/etcd` as it was originally and I've added conditional logic so that it uses the GCE persistent disk path on GCE and `/var/etcd` on Azure.

colemickens commented:
I've responded to or fixed all of the comments and pushed a new commit. Note that I didn't revert j2->jsonnet or fix the short abbreviated names in Terraform, but I did offer my comments about those suggestions above.

Overall, yes, this was more straightforward and less painful than diving into the Salt mine. I don't have a lot else to say other than that, though the thought in the back of my mind is that the Salt kube-up.sh did more than this is doing. That might just be addons though, I can't think of much else.

I have several questions though:

0. What is the expected user experience? Right now, I am entering a docker container (because I need terraform + jsonnet and they're not in my package manager), then I generate the terraform config. Then I apply the terraform config a couple times (to get around flakiness), and then I set `CLUSTER_NAME` and `MASTER_IP` and call `make set-kubeconfig`. The last step, at least, could be output by terraform since it has access to `cfg.phase1.instance_prefix` and obvious the public ip, but we still need some entrypoint script to drive the process. (And for Azure I need to help the user create the service account to use for the deployment [both in Terraform and to plumb credentials to the box because Azure doesn't have managed VM account identities]).

1. What is the plan for addons?

0. Are we taking a dependency on systemd (or ubuntu even)? (like the `systemctl` in configure-vm.sh, along with Ansible service units?)

2. Can we make Flannel work with this (the experimental overlay flag?)? Otherwise the Azure `min-turnup` will be reliant on my personal branch until k8s 1.4 with Azure CloudProvider. 

3. Is using Ansible for upgrade scenarios an explicit goal? Or, rather, are supporting upgrades with `min-turnup` an explicit goal?

4. (minor) Is `config.phase2.extra-api-sans` actually used anywhere? I thought I'd seen it at an older revision but I can't see it being used now.

And then these additional concerns:

1. Terraform is flaky for the Azure subnet. It requires retrying. Not a huge deal. I have an issue files on Terraform.

2. Terraform, a number of times, gave me failures about `kubelet.tar` missing. Which is weird because it would have existed from previous runs and been clobbered in place. Retrying bypasses this as well.

2. We should remove the root CA private key from the bundle that is uploaded to machines

colemickens commented:
Unfortunately terraform rc2 seems to have a blocking regression from rc1: https://github.com/hashicorp/terraform/issues/7248

colemickens commented:
After seeing how far ahead `kubernetes-anywhere` is (answering many of my questions above), along with @mikedanese's [`makefile` PR](https://github.com/kubernetes/kubernetes-anywhere/pull/134), I decided to move it myself.

https://github.com/colemickens/kubernetes-anywhere/commits/azure

There are some hacks in there to work around Terraform bugs and some weird bug I'm now experiencing in Docker. They're noted in comments with links to the upstream bugs.

It's not a super nice git history. I have fixed some typos and made some slight, but hopefully appropriate changes to the GCE phase1. I can split this into smaller commits if needed.

That's on top of the new, unmerged `makefile` PR. Let me know how we want to proceed here. I can go ahead and close this and open a PR on `kubernetes-anywhere` but maybe we want to have consensus here on the thing I punted from the comments.

The Azure deployment has an extra feature, it writes out a `kubeconfig` file at the end so that it's easy to connect to the cluster immediately (well not immediately, the ansible playbook docker image takes a while to pull and execute).

mikedanese commented:
Also you need to run make fmt to format the jsonnet.

mikedanese commented:
Code is gone.
